lte/gateway/python/magma/mobilityd/ip_allocator_dhcp.py

```diff
@@ -435,36 +439,39 @@ def _release_dhcp_ip(self, ip_desc: IPDesc) -> None:
         dhcp_desc = self.get_dhcp_desc_from_store(mac, vlan)
         logging.info("Releasing dhcp desc: %s", dhcp_desc)
         if dhcp_desc:
-            call_args = [
-                DHCP_HELPER_CLI,
-                "--mac", str(mac),
-                "--vlan", str(vlan),
-                "--interface", self._iface,
-                "--json",
-                "release",
-                "--ip", str(ip_desc.ip),
-                "--server-ip", str(dhcp_desc.server_ip),
-            ]
-            ret = subprocess.run(
-                call_args,
-                capture_output=True,
-                check=False,
+            self._deque_old_process(ip_desc)
+            proc = subprocess.Popen(
+                [
+                    DHCP_HELPER_CLI,
+                    "--mac", str(mac),
+                    "--vlan", str(vlan),
+                    "--interface", self._iface,
+                    "release",
+                    "--ip", str(ip_desc.ip),
+                    "--server-ip", str(dhcp_desc.server_ip),
+                ],
             )
-
-            if ret.returncode != 0:
-                call_str = " ".join(call_args)
-                logging.error(
-                    "CLI call '%s' failed with return code %s and error %s",
-                    call_str, ret.returncode, ret.stderr,
-                )
-                raise NoAvailableIPError('Failed to call dhcp_helper_cli.')
-
+            self.dhcp_helper_procs.insert(0, proc)
             key = mac.as_redis_key(vlan)
             with self.dhcp_wait:
                 del self._store.dhcp_store[key]
         else:
             LOG.error("Unallocated DHCP release for MAC: %s", mac)
 
+    def _deque_old_process(self, ip_desc: IPDesc) -> None:
+        while len(self.dhcp_helper_procs) >= self.MAX_DHCP_PROCS:
+            oldest_proc = self.dhcp_helper_procs.pop()
+            if oldest_proc.poll() is None:
+                try:
+                    oldest_proc.wait(timeout=10)
+                except subprocess.TimeoutExpired:
+                    logging.warning(
+                        "Unable to release IP %s."
+                        "Killing release process "
+                        "and moving on.", ip_desc.ip,
+                    )
+                    oldest_proc.kill()
```

Comment:
Iirc we talked about whether `.kill()` could rise an exception (an exception would prevent the follow up call to the helper CLI). Do you have an answer to that?

Reply:
Killing/terminating an already expired process does not cause an exception.
```
Python 3.8.10 (default, Jul 12 2022, 09:44:48) 
[GCC 11.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import subprocess
>>> proc = subprocess.Popen(["echo", "0"])
>>> proc.wait()
0
>>> proc.returncode
0
>>> proc.kill()
>>> proc
<subprocess.Popen object at 0x7f8d4235ccd0>
```